### PR TITLE
Make migrations run sequentially

### PR DESCRIPTION
**Pull request by evanrolfe**
I'm just making this PR as a "prototype" - something to demonstrate thats its possible that we can have the migrations run sequentially in order to generate the entire database.

All i have done is replaced the first migration [001_initial_database.rb](https://github.com/opensuse/open-build-service/blob/master/src/api/db/migrate/001_initial_database.rb#L745) with the schema dump for the obs 2.5 database.

As you can see there are some differences between structure.sql in master and between what gets generated by running the migrations sequentially. Most of those are just adding `USING BTREE` to the indices however some of the differences are non-trivial and we will need to decide which is the correct way to go for these differences if we wish to merge this PR.

### CODE REVIEW

File: src/api/db/structure.sql [hunk 1]
```diff
@@ -878,26 +878,26 @@ CREATE TABLE `repositories` (
...
-  KEY `hostsystem_id` (`hostsystem_id`),
+  UNIQUE KEY `projects_name_index` (`db_project_id`,`name`,`remote_project_name`) USING BTREE,
+  KEY `hostsystem_id` (`hostsystem_id`) USING BTREE,
+  KEY `remote_project_name_index` (`remote_project_name`) USING BTREE,
   CONSTRAINT `repositories_ibfk_1` FOREIGN KEY (`db_project_id`) REFERENCES `projects` (`id`),
   CONSTRAINT `repositories_ibfk_2` FOREIGN KEY (`hostsystem_id`) REFERENCES `repositories` (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 
 CREATE TABLE `repository_architectures` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 2]
```diff
@@ -878,26 +878,26 @@ CREATE TABLE `repositories` (
   `db_project_id` int(11) NOT NULL,
   `name` varchar(255) COLLATE utf8_bin NOT NULL,
   `remote_project_name` varchar(255) COLLATE utf8_bin NOT NULL DEFAULT '',
-  `rebuild` enum('transitive','direct','local') CHARACTER SET utf8 DEFAULT NULL,
-  `block` enum('all','local','never') CHARACTER SET utf8 DEFAULT NULL,
-  `linkedbuild` enum('off','localdep','all') CHARACTER SET utf8 DEFAULT NULL,
+  `rebuild` varchar(10) CHARACTER SET utf8 DEFAULT NULL,
+  `block` varchar(5) CHARACTER SET utf8 DEFAULT NULL,
+  `linkedbuild` varchar(8) CHARACTER SET utf8 DEFAULT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: These 3 lines are non-trivial differences, need to be checked.

File: src/api/db/structure.sql [hunk 3]
```diff
@@ -865,10 +865,10 @@ CREATE TABLE `release_targets` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `repository_id` int(11) NOT NULL,
   `target_repository_id` int(11) NOT NULL,
-  `trigger` enum('manual','allsucceeded','maintenance') DEFAULT NULL,
+  `trigger` varchar(12) DEFAULT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 4]
```diff
@@ -670,28 +670,28 @@ CREATE TABLE `messages` (
...
+  KEY `index_package_issues_on_package_id_and_issue_id` (`package_id`,`issue_id`) USING BTREE,
   CONSTRAINT `package_issues_ibfk_1` FOREIGN KEY (`package_id`) REFERENCES `packages` (`id`),
   CONSTRAINT `package_issues_ibfk_2` FOREIGN KEY (`issue_id`) REFERENCES `issues` (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
 CREATE TABLE `package_kinds` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `package_id` int(11) DEFAULT NULL,
-  `kind` enum('patchinfo','aggregate','link','channel','product') NOT NULL,
+  `kind` varchar(9) NOT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 5]
```diff
@@ -670,28 +670,28 @@ CREATE TABLE `messages` (
...
+  KEY `object` (`db_object_id`) USING BTREE,
+  KEY `user` (`user_id`) USING BTREE
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 
 CREATE TABLE `package_issues` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `package_id` int(11) NOT NULL,
   `issue_id` int(11) NOT NULL,
-  `change` enum('added','deleted','changed','kept') DEFAULT NULL,
+  `change` varchar(7) DEFAULT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 6]
```diff
@@ -616,11 +616,11 @@ CREATE TABLE `issues` (
   `owner_id` int(11) DEFAULT NULL,
   `created_at` datetime DEFAULT NULL,
   `updated_at` datetime DEFAULT NULL,
-  `state` enum('OPEN','CLOSED','UNKNOWN') CHARACTER SET utf8 DEFAULT NULL,
+  `state` varchar(7) CHARACTER SET utf8 DEFAULT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 7]
```diff
@@ -532,29 +532,29 @@ CREATE TABLE `groups` (
...
-  UNIQUE KEY `groups_roles_all_index` (`group_id`,`role_id`),
-  KEY `role_id` (`role_id`),
+  UNIQUE KEY `groups_roles_all_index` (`group_id`,`role_id`) USING BTREE,
+  KEY `role_id` (`role_id`) USING BTREE,
   CONSTRAINT `groups_roles_ibfk_1` FOREIGN KEY (`group_id`) REFERENCES `groups` (`id`),
   CONSTRAINT `groups_roles_ibfk_2` FOREIGN KEY (`role_id`) REFERENCES `roles` (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
 CREATE TABLE `groups_users` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 8]
```diff
@@ -478,27 +478,27 @@ CREATE TABLE `events` (
...
   `id` int(11) NOT NULL AUTO_INCREMENT,
-  `status` enum('enable','disable') CHARACTER SET utf8 NOT NULL,
+  `status` varchar(7) CHARACTER SET utf8 NOT NULL,
   `repo` varchar(255) CHARACTER SET utf8 DEFAULT NULL,
   `project_id` int(11) DEFAULT NULL,
   `package_id` int(11) DEFAULT NULL,
   `architecture_id` int(11) DEFAULT NULL,
   `position` int(11) NOT NULL,
-  `flag` enum('useforbuild','sourceaccess','binarydownload','debuginfo','build','publish','access','lock') CHARACTER SET utf8 NOT NULL,
+  `flag` varchar(14) CHARACTER SET utf8 NOT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql [hunk 9]
```diff
@@ -478,27 +478,27 @@ CREATE TABLE `events` (
...
+  KEY `index_events_on_eventtype` (`eventtype`) USING BTREE,
+  KEY `index_events_on_project_logged` (`project_logged`) USING BTREE,
+  KEY `index_events_on_queued` (`queued`) USING BTREE,
   KEY `index_events_on_mails_sent` (`mails_sent`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
 
 CREATE TABLE `flags` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
-  `status` enum('enable','disable') CHARACTER SET utf8 NOT NULL,
+  `status` varchar(7) CHARACTER SET utf8 NOT NULL,
```
Comment (evanrolfe) [inline src/api/db/structure.sql]: Non-trivial difference, needs to be checked.

File: src/api/db/structure.sql (hunk 7 above)
Comment (evanrolfe) [inline src/api/db/structure.sql]: Actually this is a trivial change - the id column was just moved up from a few lines below.

File: src/api/db/structure.sql (hunk 1 above)
Comment (evanrolfe) [inline src/api/db/structure.sql]: Actually this is a trivial change - the id column was just moved up from a few lines below.

**Comment by evanrolfe** [comment]
All non-trivial differences between the structure.sql files have been resolved. The only difference now is that the new structure.sql specifies BTREE as the type of index to use which is the default for mysql indices anyway.

I'm removing the "DO-NOT-MERGE" tag because this could be merged to master if we want to. I'm on holiday next week though so I won't be able to answer any questions about this PR till I'm back but basically what this PR does is:

1. Replace migration 001 with the schema dump from OBS 2.5
2. Delete all migrations which were created before OBS 2.5

**Comment by adrianschroeter** [comment]
wouldn't we need the seeds from 2.5 also at least?

**Comment by hennevogel** [comment]
I must say I really do not like this. This is so much back and forth for so little gain....

**Comment by evanrolfe** [comment]
@adrianschroeter the 2.5 seeds are already included in the first migration (although the seeds really shouldn't be handled by migrations - but this is a task for another day..)
https://github.com/evanrolfe/open-build-service/blob/72fb79ade089b5d679454812296e2368fcb63504/src/api/db/migrate/001_initial_database.rb#L766-L832

@hennevogel can you explain what you mean by "so much back and forth"?

Here are some advantages of this change:

1. It resolves the difference between whats in migrations and whats in structure.sql (see issue https://github.com/openSUSE/open-build-service/issues/2840 problem 2)
2. It allows us to easily verify what the correct structure.sql file should be - we can just run `rake db:drop db:create db:migrate` to see what structure.sql should be 
3. It allows us to squash old migrations into one single migration (either by hand or by using a gem i.e. [squasher](https://github.com/jalkoby/squasher)
4. It means the migrations no make sense, currently there is no logic to the way migrations are setup - for example this is the [first migration](https://github.com/openSUSE/open-build-service/blob/master/src/api/db/migrate/20120529150500_spaces_in_tracker_regexen.rb) but it relies on the issue_trackers table already existing.. So our migrations just start from a random point in time? 
5. It means our migrations follow the pattern that the majority of other rails applications do. This makes it easier to use third party gems with our migrations and it also means that new developers will be able to work with our code more easily because they won't have to figure out why we aren't doing things they way everybody else does them.

**Comment by hennevogel** [comment]
@evanrolfe I "just" do not agree *completely* that `db:migrate` is the way to get a *correct* structure.sql. Because migrations too often depend on the current state of your app and database. 

I have nothing against squashing old migrations once (a.k.a. **now**)...

**Comment by evanrolfe** [comment]
>  I "just" do not agree completely that db:migrate is the way to get a correct structure.sql. Because migrations too often depend on the current state of your app and database.

So the solution to out of date migrations is to squash migrations. That way we get all 7 advantages that I've listed above. Please see this blog post about a very similar problem, surely this is a better solution than what we have now?

http://naturaily.com/blog/post/how-to-remove-old-database-migrations-in-ruby-on-rails

We can still use structure.sql to load the structure into new databases (i.e. be the authoritative source) but the migrations should generate the same structure that structure.sql does.

**Comment by hennevogel** [comment]
> but the migrations should generate the same structure that structure.sql does

But they won't, that's what I'm saying. Anyway. Let's squash the migrations :-)

For that you have to rebase, drop the merge commit and fold 72fb79ade089b5d679454812296e2368fcb63504 into 6173de39a285abbbe05a5c551af7e223c03d39b9 please

**Comment by hennevogel** [comment]
@evanrolfe ping

**Comment by evanrolfe** [comment]
@hennevogel sorry for late reply. I am very keen to finish this but currently busy with the sprint so have added a DO NOT MERGE tag to this and will hopefully finish this PR sometime next week if theres time, otherwise I will create a card for this so it can be discussed in the planning for the next sprint. Is that ok or should I follow a different protocol for these kinds of tasks?

**Comment by hennevogel** [comment]
It's okay for me :-)

**Pull request merged**